### PR TITLE
Make selection based archives use relative paths

### PR DESCRIPTION
**Pull request by 0xACE**
This will create archives relative to your current working directory
rather than a full path when when creating the archive based on your
current selection.

I don't expect this to be adopted to `upstream` but issue #311 seemed to want this and I'm guessing many others expect it to be working this way.

### CODE REVIEW

**Comment by jarun** [comment]
@ath3 please review this PR. It should take care of both Linux and macOS.

File: src/nnn.c [hunk 1]
```diff
@@ -1302,7 +1302,7 @@ static void archive_selection(const char *cmd, const char *archive, const char *
 	char *buf = (char *)malloc(CMD_LEN_MAX * sizeof(char));
 	snprintf(buf, CMD_LEN_MAX,
 #ifdef __linux__
-		 "xargs -0 -a %s %s %s", g_cppath, cmd, archive);
+		"sed -ze 's|^%s/||' '%s' | xargs -0 %s %s", curpath, g_cppath, cmd, archive);
 #else
 		 "cat %s | xargs -0 -o %s %s", g_cppath, cmd, archive);
```
Comment (jarun) [inline src/nnn.c]: Please handle for macOS as well.

**Comment by 0xACE** [comment]
I uploaded a version with `MacOS` support, but I don't have a `MacOS` machine to try this with. Feel free to adjust it to your needs.

**Comment by jarun** [comment]
I don't have macOS either.

Would request whoever is here to try it out if he has access to macOS.

**Comment by serussell** [comment]
> Would request whoever is here to try it out if (s)he has access to macOS.

I get a "utility missing" error message when I try to create archives with `nnn` on OSX.

```
4231)~ % uname -a
Darwin PHLLT8526.local 18.5.0 Darwin Kernel Version 18.5.0: Mon Mar 11 20:40:32 PDT 2019; root:xnu-4903.251.3~3/RELEASE_X86_64 x86_64
4232)~ % which zip
/usr/bin/zip
4233)~ % which tar
/usr/bin/tar
4234)~ % zip -v
Copyright (c) 1990-2008 Info-ZIP - Type 'zip "-L"' for software license.
This is Zip 3.0 (July 5th 2008), by Info-ZIP.
...
```

**Comment by jarun** [comment]
Update to master.

**Comment by 0xACE** [comment]
>I get a "utility missing" error message when I try to create archives with nnn on OSX.

This patch only introduces `tr` and `sed`. `sed` is the only new reference in the entire codebase, so if it's missing a utility that would be it. Or maybe you are running a old version as jarun said...

**Comment by serussell** [comment]
Just to confirm: this morning, I did a fresh checkout of the branch "relativearchives" from https://github.com/0xACE/nnn; I'm on commit 2baa42b877ca0bc338a1852397ca3a0aa7c30b79.  That's what was the request for OSX testing, right?  And:

```
4245)~/nnn % which tr
/usr/bin/tr
4246)~/nnn % which sed
/usr/bin/sed
```

**Comment by serussell** [comment]
*<sigh>* there was another `nnn` in a different path on OSX, so PEBKAC.

Confirming the patch works as expected on OSX.  The only oddity is that I created `Foo.zip`, and that's the name of the file that got created, but what *actually* got created was a tarball:

```
4257)Documents/architecture % file Foo.zip
Foo.zip: POSIX tar archive
```

**Comment by 0xACE** [comment]
> Confirming the patch works as expected on OSX. The only oddity is that I named the archive `Foo.zip`, and that's what got created, but the file type that was created wasn't a zip, but a tarball:
> 
> ```
> 4257)Documents/architecture % file Foo.zip
> Foo.zip: POSIX tar archive
> ```

I would argue the problem resides from where `archive_selection()` is called.

I'm guessing 8e00d74259f98d707e9ba04d982c34d4da66a584 with `get_archive_cmd()` is the culprit where `else if (getutil(utils[BSDTAR]))` should be moved further down the if-else statements...

**Comment by jarun** [comment]
I can reproduce this. And that's strange! The [man page of bsdtar](https://www.freebsd.org/cgi/man.cgi?query=bsdtar&sektion=1&manpath=freebsd-release-ports) description says:

`... can create tar, pax,	cpio, ar, zip, 7-zip, and shar archives.`

So why would it create an archive with the extension `zip` as `tar`?

**Comment by jarun** [comment]
Ahh, we need to use the option `-a`!

**Comment by jarun** [comment]
Thanks @0xACE! We now have enough time till the next release to get it tested thoroughly.

**Comment by 0xACE** [comment]
Yeah, I'm not stressed.

But one concern I have is shouldn't the the `#if define(__APPLE__)` just be the `#else` statement so non  `__APPLE__` ?

I'm not familiar with github interface. But I wrote comments on the affected code above. I figured it didn't ping you or you didn't mind either way.

**Comment by jarun** [comment]
> But one concern I have is shouldn't the the #if define(__APPLE__) just be the #else statement so non __APPLE__ ?

Mostly it would be BSD people and they don't mind non-issues.

**Comment by 0xACE** [comment]
>Mostly it would be BSD people and they don't mind non-issues.

Haha, great. I'll maintain a local copy that uses `#else` and annotate why upstream is different

**Comment by jarun** [comment]
I see that FreeBSD, OpenBSD and NetBSD have `sed`, `cat` and `tr` so I think it's safe to include those along with macOS. I am making the case #else and let's see if someone finds an issue. As I mentioned we have time till next release.

**Comment by jarun** [comment]
Pushed at commit 0647c75cd8c75f5377b24b1be2456d1f71f649b3.

**Comment by 0xACE** [comment]
> I see that FreeBSD, OpenBSD and NetBSD have `sed`, `cat` and `tr` so I think it's safe to include those along with macOS.

Yes. To clarify: `GNU sed` has the `-z` flag to seperate by lines by `\0`. Howver appels `sed` doesn't seem to have that feature which is why the apple version is using `tr` instead... Essentially if `BSD` also has `sed -z` then the first version would suffice for them...

>I am making the case #else and let's see if someone finds an issue. As I mentioned we have time till next release.

Great! I'll drop my local copy

**Comment by jarun** [comment]
Looks like none of the BSDs have `-z`. Generally they are similar to macOS.

**Comment by jarun** [comment]
I don't understand this: https://wiki.archlinux.org/index.php?title=Nnn&type=revision&diff=578979&oldid=578937

minimalist and lightweight are characteristics of `nnn` and not ads. I would have loved to discuss this with the editor but i don't know the ways of the Arch wiki. Can we do somehting here? I totally believe it's a wrong reversal.

**Comment by 0xACE** [comment]
I'm not sure. I wouldn't push my luck. I'm too busy to deal with it this until the later half of september. But I guess it could be discussed on that wiki page... iirc I have seen other tools advertise as "minimalistic and lightweight"

**Comment by jarun** [comment]
I think we can ignore it. Our performance page says it all and I am pushing more content to our wiki from the homepage.

I must say you were right about the proposal to move the hacking nnn page to arch wiki.

**Pull request merged**